services/brig/test/integration/API/User/Auth.hs

```diff
@@ -968,6 +974,157 @@ getAndTestDBSupersededCookieAndItsValidSuccessor config b n = do
+    ZAuth.userTokenClient @ZAuth.User token @?= Just (clientId cl)
+    assertSaneAccessToken now (userId u) (decodeToken' @ZAuth.Access r)
+    ZAuth.accessTokenClient @ZAuth.Access atoken @?= Just (clientId cl)
+
+testAccessWithIncorrectClientId :: Brig -> Http ()
+testAccessWithIncorrectClientId brig = do
+  u <- randomUser brig
+  rs <-
+    login
+      brig
+      ( emailLogin
+          (fromJust (userEmail u))
+          defPassword
+          (Just "nexus1")
+      )
+      PersistentCookie
+      <!! const 200 === statusCode
+  let c = decodeCookie rs
+  addClient
+    brig
+    (userId u)
+    (defNewClient PermanentClientType [] (Imports.head someLastPrekeys))
+    !!! const 201 === statusCode
+  post
+    ( unversioned
+        . brig
+        . path "/access"
+        . queryItem "client_id" "beef"
+        . cookie c
+    )
+    !!! const 403 === statusCode
+
+testAccessWithExistingClientId :: Brig -> Http ()
+testAccessWithExistingClientId brig = do
+  u <- randomUser brig
+  rs <-
+    login
+      brig
+      ( emailLogin
+          (fromJust (userEmail u))
+          defPassword
+          (Just "nexus1")
+      )
+      PersistentCookie
+      <!! const 200 === statusCode
+  let c0 = decodeCookie rs
+  cl <-
+    responseJsonError
+      =<< addClient
+        brig
+        (userId u)
+        (defNewClient PermanentClientType [] (Imports.head someLastPrekeys))
+      <!! const 201 === statusCode
+  now <- liftIO getCurrentTime
+
+  -- access with client ID first
+  c1 <- do
+    r <-
+      post
+        ( unversioned
+            . brig
+            . path "/access"
+            . queryItem "client_id" (toByteString' (clientId cl))
+            . cookie c0
+        )
+        <!! const 200 === statusCode
+    pure (decodeCookie r)
+
+  -- now access again with no client ID
+  c2 <- do
+    r <-
+      post
+        ( unversioned
+            . brig
+            . path "/access"
+            . cookie c1
+        )
+        <!! const 200 === statusCode
+    liftIO $ do
+      let ck = decodeCookie r
+          Just token = fromByteString (cookie_value ck)
+          atoken = decodeToken' @ZAuth.Access r
+      assertSanePersistentCookie @ZAuth.User ck
+      ZAuth.userTokenClient @ZAuth.User token @?= Just (clientId cl)
+      assertSaneAccessToken now (userId u) (decodeToken' @ZAuth.Access r)
+      ZAuth.accessTokenClient @ZAuth.Access atoken @?= Just (clientId cl)
+    pure (decodeCookie r)
+
+  -- now access with a different client ID
+  do
+    cl2 <-
+      responseJsonError
+        =<< addClient
+          brig
+          (userId u)
+          (defNewClient PermanentClientType [] (someLastPrekeys !! 1))
+        <!! const 201 === statusCode
+    r <-
+      post
+        ( unversioned
+            . brig
+            . path "/access"
+            . queryItem "client_id" (toByteString' (clientId cl2))
+            . cookie c2
+        )
+        <!! const 200 === statusCode
```

Comment:
I suppose this is fine, but would it make more sense to return an error code in this case, i.e., when the provided client ID is different from what was initially provided?

Reply:
I guess so, I was hesitant about it because of the awkward monad stack that is needed to get that. This is because `MonadLogger` does not have an instance for `ExceptT`. But it's not the end of the world.